File: crates/kn9t-tui/tests/policy_ui.lua

-- Policy plugin UI
local V = { cursor = 0, adding = false, input = "" }

function on_state(s)
    V.mode = s.mode or "normal"
    V.grants = s.grants or {}
    V.recent = s.recent or {}
    -- Don't override cursor/adding/input from state - those are local UI state
end

-- Exact keys first: the host matches on_key before on_text, so a shortcut aimed
-- at a printable character must decline while a grant is being typed. That
-- leaves the single on_text handler below as the only writer of the input
-- buffer. Actions that persist (mode, grants) go to Python via kn9t.notify;
-- navigation stays local in V.

kn9t.on_key("m", function()
    if V.adding then return false end
    kn9t.notify({ event = "cycle_mode" })
    return true
end)

kn9t.on_key("a", function()
    if V.adding then return false end
    V.adding = true
    V.input = ""
    return true
end)

-- `d` and `x` are the same action.
local function delete_grant()
    if V.cursor >= 0 and V.cursor < #V.grants then
        kn9t.notify({ event = "delete_grant", index = V.cursor })
    end
    return true
end

kn9t.on_key("d", function() if V.adding then return false end return delete_grant() end)
kn9t.on_key("x", function() if V.adding then return false end return delete_grant() end)

local function move(delta)
    local n = #V.grants
    if n == 0 then return end
    V.cursor = math.max(0, math.min(n - 1, V.cursor + delta))
end

kn9t.on_key("j", function() if V.adding then return false end move(1) return true end)
kn9t.on_key("k", function() if V.adding then return false end move(-1) return true end)
kn9t.on_key("Down", function() if V.adding then return false end move(1) return true end)
kn9t.on_key("Up", function() if V.adding then return false end move(-1) return true end)

kn9t.on_key("Escape", function()
    -- Clear a half-typed grant, then let Esc release focus.
    if V.adding then
        V.adding = false
        V.input = ""
    end
    return false
end)

kn9t.on_key("Enter", function()
    if not V.adding or V.input == "" then return false end
    kn9t.notify({ event = "add_grant", pattern = V.input })
    V.adding = false
    V.input = ""
    return true
end)

kn9t.on_key("Backspace", function()
    if not V.adding then return false end
    V.input = string.sub(V.input, 1, -2)
    return true
end)

-- One handler for every printable character, Space included. It owns the
-- keystroke only while a grant is being typed; otherwise the character falls
-- through to the shortcuts above or to the host.
kn9t.on_text(function(ch)
    if not V.adding then return false end
    V.input = V.input .. ch
    return true
end)

function render(s)
    on_state(s)
    local out = {}
    local C = {
        accent = "cyan",
        dim = "darkgray",
        normal = "green",
        yolo = "yellow",
        ask_all = "magenta",
        allow = "green",
        deny = "lightred",
        ask = "yellow",
    }
    
    -- Mode selector
    local mode_spans = {}
    table.insert(mode_spans, { text = "Mode: ", fg = C.dim })
    for _, m in ipairs({"normal", "yolo", "ask_all"}) do
        if V.mode == m then
            table.insert(mode_spans, { text = "[", fg = C[m] })
            table.insert(mode_spans, { text = m, fg = C[m], bold = true })
            table.insert(mode_spans, { text = "] ", fg = C[m] })
        else
            table.insert(mode_spans, { text = m .. " ", fg = C.dim })
        end
    end
    table.insert(out, { type = "text", spans = mode_spans, size = { fixed = 1 } })
    
    -- Separator
    table.insert(out, { type = "text", content = string.rep("─", 40), fg = C.dim, size = { fixed = 1 } })
    
    -- Recent decisions
    table.insert(out, { type = "text", content = "Recent:", fg = C.dim, size = { fixed = 1 } })
    if #V.recent == 0 then
        table.insert(out, { type = "text", content = "  (none)", fg = C.dim, size = { fixed = 1 } })
    else
        for i = #V.recent, 1, -1 do
            local d = V.recent[i]
            local icon = "✓"
            local col = C.allow
            if d.result == "deny" then icon = "✗"; col = C.deny
            elseif d.result == "ask" then icon = "?"; col = C.ask end
            local spans = {
                { text = icon .. " ", fg = col },
                { text = d.cmd, fg = "white" },
            }
            if d.reason ~= "" then
                table.insert(spans, { text = " (" .. d.reason .. ")", fg = C.dim })
            end
            table.insert(out, { type = "text", spans = spans, size = { fixed = 1 } })
        end
    end
    
    -- Separator
    table.insert(out, { type = "text", content = string.rep("─", 40), fg = C.dim, size = { fixed = 1 } })
    
    -- Grants list
    table.insert(out, { type = "text", content = "Grants (always allow):", fg = C.dim, size = { fixed = 1 } })
    if #V.grants == 0 and not V.adding then
        table.insert(out, { type = "text", content = "  (none) - press 'a' to add", fg = C.dim, size = { fixed = 1 } })
    else
        for i, g in ipairs(V.grants) do
            local prefix = (i - 1 == V.cursor) and "> " or "  "
            local fg = (i - 1 == V.cursor) and C.accent or "white"
            table.insert(out, { type = "text", content = prefix .. g, fg = fg, size = { fixed = 1 } })
        end
    end
    
    -- Input line for adding
    if V.adding then
        local spans = {
            { text = "  + ", fg = C.accent },
            { text = V.input, fg = "white" },
            { text = "█", fg = C.accent },
        }
        table.insert(out, { type = "text", spans = spans, size = { fixed = 1 } })
    end
    
    -- Spacer
    table.insert(out, { type = "spacer", size = { flex = 1 } })
    
    -- Help bar
    local help_spans = {}
    if V.adding then
        table.insert(help_spans, { text = "[Enter]", fg = C.accent })
        table.insert(help_spans, { text = " save  ", fg = C.dim })
        table.insert(help_spans, { text = "[Bksp]", fg = C.accent })
        table.insert(help_spans, { text = " clear", fg = C.dim })
    else
        table.insert(help_spans, { text = "[m]", fg = C.accent })
        table.insert(help_spans, { text = " mode  ", fg = C.dim })
        table.insert(help_spans, { text = "[a]", fg = C.accent })
        table.insert(help_spans, { text = " add  ", fg = C.dim })
        table.insert(help_spans, { text = "[d]", fg = C.accent })
        table.insert(help_spans, { text = " del  ", fg = C.dim })
        table.insert(help_spans, { text = "[↑↓]", fg = C.accent })
        table.insert(help_spans, { text = " nav", fg = C.dim })
    end
    table.insert(out, { type = "text", spans = help_spans, size = { fixed = 1 } })
    
    return { type = "split", direction = "vertical", children = out }
end

File: crates/kn9t-tui/tests/policy_ui.rs
//! The `kn9t-policy` plugin's Lua UI, extracted verbatim from its Python source
//! (`scripts/extract_policy_lua.py`), so this is not a hand-written approximation.
//! The plugin lives in the kn9t-plugins repo —
//! https://github.com/96Ems/kn9t-plugins/tree/main/kn9t-policy.
//!
//! These tests lock the migration to the generic primitives: the grant input is
//! driven by `kn9t.on_text`, not by a per-character `on_key` loop, and the
//! shortcuts must still fire when nothing is being typed.

use kn9t_tui::lua::plugin_ui::PluginEffect;
use kn9t_tui::lua::widgets::Widget;
use kn9t_tui::lua::LuaRuntime;
use kn9t_tui::reducer::{PluginLuaOp, PluginPlacement};
use serde_json::json;

/// The Lua the real plugin sends over the wire.
const POLICY_LUA: &str = include_str!("policy_ui.lua");

const PLUGIN: &str = "kn9t-policy";

fn runtime_with_policy(state: serde_json::Value) -> LuaRuntime {
    let rt = LuaRuntime::new().unwrap();
    rt.load_builtin();
    rt.apply_plugin_lua_op(&PluginLuaOp::Register {
        plugin: PLUGIN.into(),
        source: POLICY_LUA.into(),
        placement: PluginPlacement {
            zone: Some("main".into()),
            ..Default::default()
        },
    });
    rt.apply_plugin_lua_op(&PluginLuaOp::SetState {
        plugin: PLUGIN.into(),
        state,
    });
    rt
}

fn state(grants: serde_json::Value) -> serde_json::Value {
    json!({ "mode": "normal", "grants": grants, "recent": [] })
}

/// Collect every text string in a widget tree, so assertions do not depend on
/// the exact nesting the plugin chose.
fn all_text(w: &Widget, out: &mut Vec<String>) {
    match w {
        Widget::Text { .. } => out.extend(w.text()),
        Widget::List { items, .. } => out.extend(
            items
                .iter()
                .map(|spans| spans.iter().map(|s| s.text.as_str()).collect::<String>()),
        ),
        Widget::Box { title, child, .. } => {
            if let Some(t) = title {
                out.push(t.clone());
            }
            if let Some(c) = child {
                all_text(c, out);
            }
        }
        Widget::Float { child: Some(c), .. } => all_text(c, out),
        Widget::Split { children, .. } => {
            for c in children {
                all_text(c, out);
            }
        }
        _ => {}
    }
}

fn texts(rt: &LuaRuntime) -> Vec<String> {
    let w = rt.build_plugin_view(PLUGIN).expect("policy Lua must build");
    let mut out = Vec::new();
    all_text(&w, &mut out);
    out
}

/// The shipped Lua must load — a syntax error would fail before a user runs it.
#[test]
fn shipped_lua_loads_and_renders() {
    let rt = runtime_with_policy(state(json!([])));
    assert!(
        rt.build_plugin_view(PLUGIN).is_ok(),
        "the plugin's shipped Lua must load"
    );
}

/// Typing goes through the single `kn9t.on_text` handler: `a` opens the input,
/// then characters accumulate, and the rendered input line shows them.
#[test]
fn typed_characters_land_in_the_grant_input() {
    let rt = runtime_with_policy(state(json!([])));
    assert!(rt.dispatch_plugin_key(PLUGIN, "a"), "`a` must start adding");
    assert!(
        rt.dispatch_plugin_text(PLUGIN, "h"),
        "on_text consumes while adding"
    );
    assert!(
        rt.dispatch_plugin_text(PLUGIN, "i"),
        "on_text consumes while adding"
    );

    let t = texts(&rt);
    assert!(
        t.iter().any(|s| s.contains("hi")),
        "typed text must reach the input line, got {t:?}"
    );
}

/// Outside the input mode the text handler declines, so its characters fall
/// through to the shortcuts (`on_key` runs first) or the host.
#[test]
fn text_falls_through_when_not_adding() {
    let rt = runtime_with_policy(state(json!([])));
    assert!(rt.plugin_has_text(PLUGIN), "the view must register on_text");
    assert!(
        !rt.dispatch_plugin_text(PLUGIN, "z"),
        "an unbound character must fall through"
    );
}

/// Enter while adding notifies the backend with exactly what was typed.
#[test]
fn enter_notifies_with_the_typed_grant() {
    let rt = runtime_with_policy(state(json!([])));
    assert!(rt.dispatch_plugin_key(PLUGIN, "a"));
    for ch in ["c", "u", "r", "l", " ", "*"] {
        assert!(
            rt.dispatch_plugin_text(PLUGIN, ch),
            "`{ch}` must be consumed"
        );
    }
    assert!(
        rt.dispatch_plugin_key(PLUGIN, "Enter"),
        "Enter must consume"
    );

    let effects = rt.drain_plugin_effects();
    assert_eq!(effects.len(), 1, "expected one effect, got {effects:?}");
    match &effects[0] {
        PluginEffect::NotifyPlugin {
            plugin,
            event,
            data,
        } => {
            assert_eq!(plugin, PLUGIN);
            assert_eq!(event, "add_grant");
            assert_eq!(data.get("pattern").and_then(|v| v.as_str()), Some("curl *"));
        }
        other => panic!("expected NotifyPlugin, got {other:?}"),
    }
}

/// Esc clears a half-typed grant and then falls through, so it still releases
/// focus (it must not trap the user in the panel).
#[test]
fn escape_clears_the_input_and_releases_focus() {
    let rt = runtime_with_policy(state(json!([])));
    assert!(rt.dispatch_plugin_key(PLUGIN, "a"));
    assert!(rt.dispatch_plugin_text(PLUGIN, "x"));
    assert!(
        texts(&rt).iter().any(|s| s.contains("x")),
        "the character must be visible before Esc"
    );

    assert!(
        !rt.dispatch_plugin_key(PLUGIN, "Escape"),
        "Esc must fall through so the host releases focus"
    );
    assert!(
        !texts(&rt).iter().any(|s| s.contains('x')),
        "Esc must clear the half-typed grant"
    );
}

/// Navigation is local: `j` moves the selection down the grants list.
#[test]
fn j_moves_the_grant_selection() {
    let rt = runtime_with_policy(state(json!(["alpha", "beta"])));
    // Render once so `on_state` populates V.grants.
    let _ = texts(&rt);
    assert!(rt.dispatch_plugin_key(PLUGIN, "j"), "`j` must move");
    let t = texts(&rt);
    assert!(
        t.iter().any(|s| s.contains("> beta")),
        "`j` must select the second grant, got {t:?}"
    );
}
